procedures/statement_performance_analyzer.sql

```diff
@@ -0,0 +1,792 @@
+            DROP TEMPORARY TABLE IF EXISTS tmp_digests;
+        END IF;
+        SET @sys.tmp.SQL = REPLACE(REPLACE(v_digest_table_template, '%{TEMPORARY}', 'TEMPORARY '), '%{TABLE_NAME}', 'tmp_digests');
+        IF (@sys.debug = 'ON') THEN
+            SELECT @sys.tmp.SQL AS 'Debug';
+        END IF;
+        PREPARE stmt_create_table FROM @sys.tmp.SQL;
+        EXECUTE stmt_create_table;
+        DEALLOCATE PREPARE stmt_create_table;
+
+        SET @sys.tmp.SQL = CONCAT('INSERT INTO tmp_digests SELECT * FROM ',
+                                  IF(in_table IS NULL OR in_action = 'save', 'performance_schema.events_statements_summary_by_digest', v_quoted_table));
+        IF (@sys.debug = 'ON') THEN
+            SELECT @sys.tmp.SQL AS 'Debug';
+        END IF;
+        PREPARE stmt_insert FROM @sys.tmp.SQL;
+        EXECUTE stmt_insert;
+        DEALLOCATE PREPARE stmt_insert;
+    END IF;
+
+    -- Go through the remaining actions
+    IF (in_action IN ('create_table', 'create_tmp')) THEN
+        IF (in_action = 'create_table') THEN
+            SET @sys.tmp.SQL = REPLACE(REPLACE(v_digest_table_template, '%{TEMPORARY}', ''), '%{TABLE_NAME}', v_quoted_table);
+        ELSE
+            SET @sys.tmp.SQL = REPLACE(REPLACE(v_digest_table_template, '%{TEMPORARY}', 'TEMPORARY '), '%{TABLE_NAME}', v_quoted_table);
+        END IF;
+        IF (@sys.debug = 'ON') THEN
+            SELECT @sys.tmp.SQL AS 'Debug';
+        END IF;
+        PREPARE stmt_create_table FROM @sys.tmp.SQL;
+        EXECUTE stmt_create_table;
+        DEALLOCATE PREPARE stmt_create_table;
+
+    ELSEIF (in_action = 'save') THEN
+        SET @sys.tmp.SQL = CONCAT('DELETE FROM ', v_quoted_table);
+        IF (@sys.debug = 'ON') THEN
+            SELECT @sys.tmp.SQL AS 'Debug';
+        END IF;
+        PREPARE stmt_delete FROM @sys.tmp.SQL;
+        EXECUTE stmt_delete;
+        DEALLOCATE PREPARE stmt_delete;
+        
+        SET @sys.tmp.SQL = CONCAT('INSERT INTO ', v_quoted_table, ' SELECT * FROM tmp_digests');
+        IF (@sys.debug = 'ON') THEN
+            SELECT @sys.tmp.SQL AS 'Debug';
+        END IF;
+        PREPARE stmt_insert FROM @sys.tmp.SQL;
+        EXECUTE stmt_insert;
+        DEALLOCATE PREPARE stmt_insert;
+
+    ELSEIF (in_action IN ('overall', 'delta')) THEN
+        -- These are almost the same - for delta calculate the delta in tmp_digests_delta and use that instead of tmp_digests.
+        -- And overall allows overriding the table to use.
+        IF (in_action = 'overall') THEN
+            IF (in_table IS NULL) THEN
+                SET v_digests_table = 'tmp_digests';
+            ELSE
+                SET v_digests_table = v_quoted_table;
+            END IF;
+        ELSE
+            SET v_digests_table = 'tmp_digests_delta';
+            DROP TEMPORARY TABLE IF EXISTS tmp_digests_delta;
+            CREATE TEMPORARY TABLE tmp_digests_delta LIKE tmp_digests;
+            SET @sys.tmp.SQL = CONCAT('INSERT INTO tmp_digests_delta
```

Comment:
For larger statements like this, the execute_prepared_stmt procedure could still take this user variable (just for readability really), and still just allocate that to another internal user variable.

It also shows that the size of the query could be pretty large too, so make sure that the execute_prepared_stmt procedure *also* nulls out the temp user variable after it's been used.